File: leetcode/120_Triangle.py
from typing import List
from unittest import TestCase

class Solution:
    def minimumTotal(self, triangle: List[List[int]]) -> int:
        n = len(triangle)
        dp = [0] * n
        dp[0] = triangle[0][0]
        for i in range(1, n):
            dp[i] = dp[i - 1] + triangle[i][i]
            for j in range(i - 1, 0, -1):
                dp[j] = triangle[i][j]  + (dp[j - 1]  if dp[j - 1] < dp[j] else dp[j])
            dp[0] += triangle[i][0]
        return min(dp)

class MinimumTotalTest(TestCase):
    def test_1(self) -> None:
        self.assertEqual(Solution().minimumTotal([[2],[3,4],[6,5,7],[4,1,8,3]]), 11)

    def test_2(self) -> None:
        self.assertEqual(Solution().minimumTotal([[-10]]), -10)

File: leetcode/120_Triangle.rs
struct Solution {}

impl Solution {
    pub fn minimum_total(triangle: Vec<Vec<i32>>) -> i32 {
        let mut triangle = triangle;
        for i in 1..triangle.len() {
            let n = triangle[i].len();
            triangle[i][0] += triangle[i - 1][0];
            triangle[i][n - 1] += triangle[i - 1][n - 2];
            for j in 1..n - 1 {
                triangle[i][j] += triangle[i - 1][j].min(triangle[i - 1][j - 1]);
            }
        }
        *triangle.last().unwrap().iter().min().unwrap()
    }
}


#[test] 
fn test_1() {
    assert_eq!(11, Solution::minimum_total(vec![vec![2],vec![3,4],vec![6,5,7],vec![4,1,8,3]]));
}

#[test] 
fn test_2() {
    assert_eq!(-10, Solution::minimum_total(vec![vec![-10]]));
}